lib/trie/decode/key.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2021 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package decode
+
+import (
+	"errors"
+	"fmt"
+	"io"
+)
+
+const maxPartialKeySize = ^uint16(0)
+
+var (
+	ErrPartialKeyTooBig = errors.New("partial key length cannot be larger than or equal to 2^16")
+	ErrReadKeyLength    = errors.New("cannot read key length")
+	ErrReadKeyData      = errors.New("cannot read key data")
+)
+
+// Key decodes a key from a reader.
+func Key(reader io.Reader, keyLength byte) (b []byte, err error) {
+	publicKeyLength := int(keyLength)
+
+	if keyLength == 0x3f {
+		// partial key longer than 63, read next bytes for rest of pk len
+		for {
+			nextKeyLen, err := ReadNextByte(reader)
+			if err != nil {
+				return nil, fmt.Errorf("%w: %s", ErrReadKeyLength, err)
+			}
+			publicKeyLength += int(nextKeyLen)
+
+			if nextKeyLen < 0xff {
+				break
+			}
+
+			if publicKeyLength >= int(maxPartialKeySize) {
+				return nil, fmt.Errorf("%w: %d",
+					ErrPartialKeyTooBig, publicKeyLength)
+			}
+		}
+	}
+
+	if publicKeyLength == 0 {
+		return []byte{}, nil
+	}
+
+	key := make([]byte, publicKeyLength/2+publicKeyLength%2)
+	n, err := reader.Read(key)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", ErrReadKeyData, err)
+	} else if n != len(key) {
+		return nil, fmt.Errorf("%w: read %d bytes instead of %d",
+			ErrReadKeyData, n, len(key))
+	}
+
+	return KeyLEToNibbles(key)[publicKeyLength%2:], nil
+}
+
+// KeyLEToNibbles converts a Little Endian byte slice into nibbles.
+// It assumes bytes are already in Little Endian and does not rearrange nibbles.
+func KeyLEToNibbles(in []byte) (nibbles []byte) {
```

Comment:
I feel like there should be some sort of `Key` type, and this function should just be a receiver function on that type.
`Key.Nibbles() ([]byte)`.

Reply:
I started working on it in a branch, but it's a lot more changes than I expected. I will pause it for now but taking note to do it at a later time eventually.